framework/web/UrlNormalizer.php

```diff
@@ -0,0 +1,125 @@
+     * @see https://en.wikipedia.org/wiki/HTTP_404
+     */
+    const ACTION_NOT_FOUND = 404;
+
+    /**
+     * @var boolean whether slashes should be collapsed, for example `site///index` will be
+     * converted into `site/index`
+     */
+    public $collapseSlashes = true;
+    /**
+     * @var boolean whether trailing slash should be normalized according to the suffix settings
+     * of the rule
+     */
+    public $normalizeTrailingSlash = true;
+    /**
+     * @var integer|callable|null action to perform during route normalization.
+     * Available options are:
+     * - `null` - no special action will be performed
+     * - `301` - the request should be redirected to the normalized URL using
+     * permanent redirection
+     * - `302` - the request should be redirected to the normalized URL using
+     * temporary redirection
+     * - `404` - [[NotFoundHttpException]] will be thrown
+     * - `callable` - custom user callback, for example:
+     *
+     *   ```php
+     *   function ($route, $normalizer) {
+     *       // use custom action for redirections
+     *       $route[1]['oldRoute'] = $route[0];
+     *       $route[0] = 'site/redirect';
+     *       return $route;
+     *   }
+     *   ```
+     */
+    public $action = self::ACTION_REDIRECT_PERMANENT;
+
+    /**
+     * Performs normalization action for specified route.
+     * @param $route route for normalization
+     * @return array normalized route
+     * @throws InvalidConfigException if invalid normalization action is used
+     * @throws UrlNormalizerRedirectException if normalization requires redirection
+     */
+    public function normalizeRoute($route)
+    {
+        if ($this->action === null) {
+            return $route;
+        } elseif ($this->action === static::ACTION_REDIRECT_PERMANENT
+                || $this->action === static::ACTION_REDIRECT_TEMPORARY) {
+            throw new UrlNormalizerRedirectException($route[0], $route[1], $this->action);
+        } elseif ($this->action === static::ACTION_NOT_FOUND) {
+            throw new NotFoundHttpException(Yii::t('yii', 'Page not found.'));
+        } elseif (is_callable($this->action)) {
+            return call_user_func($this->action, $route, $this);
+        } else {
+            throw new InvalidConfigException('Invalid normalizer action.');
+        }
+    }
+
+    /**
+     * Normalizes specified pathInfo.
+     * @param $pathInfo string pathInfo for normalization
+     * @param $normalized boolean if specified, this variable will be set to true if pathInfo
+     * was changed during normalization
+     * @return string normalized pathInfo
+     */
+    public function normalizePathInfo($pathInfo, $suffix, &$normalized = false)
+    {
+        if (empty($pathInfo)) {
+            return $pathInfo;
+        }
+
+        $sourcePathInfo = $pathInfo;
+        if ($this->collapseSlashes) {
+            $pathInfo = ltrim(preg_replace('#/{2,}#', '/', $pathInfo), '/');
```

Comment:
What are the benefits of this? Moving this to private method only make it harder to extend.